neo4j/internal/bolt/bolt4.go

```diff
@@ -983,3 +1007,26 @@ func (b *bolt4) initializeReadTimeoutHint(hints map[string]interface{}) {
 	}
 	b.in.connReadTimeout = time.Duration(readTimeout) * time.Second
 }
+
+func routeDatabaseContext(database string, impersonatedUser string) map[string]string {
+	dbContext := make(map[string]string)
+	if database != db.DefaultDatabase {
+		dbContext["db"] = database
+	}
+	if impersonatedUser != "" {
+		dbContext["imp_user"] = impersonatedUser
+	}
+	return dbContext
+}
+
+func (b *bolt4) setImpersonationConfig(config db.TxConfig, tx *internalTx4) {
+	tx.impersonatedUser = config.ImpersonatedUser
+	// b.databaseName may have been overwritten by the session
+	// else it's the same as the impersonated user's home DB
+	// in the latter case, do not include the database since it might be outdated
+	// indeed, the home DB might have switched between the routing table retrieval
+	// and the execution of BEGIN, so let the server resolve the home DB again instead
+	if tx.databaseName == b.impersonatedUserHomeDb {
+		tx.databaseName = ""
+	}
```

Comment:
In both places, where this function is called, `tx.databaseName` has been set to `b.databaseName`. So this could as well be `b.databaseName == b.impersonatedUserHomeDb`. Which seems to be always true according to https://github.com/neo4j/neo4j-go-driver/pull/269/files#diff-123e0e61642d0d876f8df56b27c5f4a262baa6ef45ba4ef003c337fdb6f55aceR903-R904

Reply:
It's actually not always true, since `SelectDatabase` might be called by the database provided by SessionConfig (see https://github.com/neo4j/neo4j-go-driver/blob/5a14c7024ca3203d89d54ae34bbfbc2886249401/neo4j/session.go#L372-L379), hence the check